cpp/src/join/joining.cu

```diff
@@ -242,149 +210,175 @@ gdf_error join_call( int num_cols, gdf_column **leftcol, gdf_column **rightcol,
+        CUDF_EXPECTS(gdf_error_code == GDF_SUCCESS, "GDF Error");
         break;
       }
     case GDF_SORT:
       {
         // Sort based joins only support single column joins
         if(1 == num_cols)
         {
-          gdf_error_code =  sort_join<join_type, output_index_type>(leftcol[0], rightcol[0], left_result, right_result);
+          gdf_error_code =  sort_join<join_type, output_index_type>(const_cast <gdf_column*> (leftcols.get_column(0)), 
+                            const_cast <gdf_column*> (rightcols.get_column(0)), left_result, right_result);
+          CUDF_EXPECTS(gdf_error_code == GDF_SUCCESS, "GDF Error");
         }
         else
         {
-          gdf_error_code =  GDF_JOIN_TOO_MANY_COLUMNS;
+          CUDF_EXPECTS(false, "Too many columns to join");
         }
 
         break;
       }
     default:
-      gdf_error_code =  GDF_UNSUPPORTED_METHOD;
+      CUDF_EXPECTS(false, "Unsupported Method");
   }
 
   POP_RANGE();
-
-  return gdf_error_code;
 }
 
-
-
 template <JoinType join_type, typename index_type>
-gdf_error construct_join_output_df(
-        std::vector<gdf_column*>& ljoincol,
-        std::vector<gdf_column*>& rjoincol,
-        gdf_column **left_cols, 
-        int num_left_cols,
-        int left_join_cols[],
-        gdf_column **right_cols,
-        int num_right_cols,
-        int right_join_cols[],
-        int num_cols_to_join,
-        int result_num_cols,
-        gdf_column ** result_cols,
+std::pair<cudf::table, cudf::table> construct_join_output_df(
+        cudf::table const& ljoincols,
+        cudf::table const& rjoincols,
+        cudf::table const& left_cols, 
+        cudf::table const& right_cols,
+        cudf::table const& join_cols,
+        cudf::table const& merging_join_cols,
         gdf_column * left_indices,
         gdf_column * right_indices) {
 
 
-  PUSH_RANGE("LIBGDF_JOIN_OUTPUT", JOIN_COLOR);
+    PUSH_RANGE("LIBGDF_JOIN_OUTPUT", JOIN_COLOR);
     //create left and right input table with columns not joined on
+    int num_left_cols = left_cols.num_columns();
+    int num_right_cols = right_cols.num_columns();
+    int num_cols_to_join = (join_cols.num_columns() > 0 ? join_cols.get_column(0)->size: 0);
+    int num_merging_join_cols = (merging_join_cols.num_columns() > 0 ? merging_join_cols.get_column(0)->size: 0);
+
+    std::vector<int> left_j_cols (num_cols_to_join);
+    std::vector<int> l_common_name_join_ind (num_merging_join_cols);
+    std::vector<int> r_common_name_join_ind (num_merging_join_cols);
+
+    if (num_cols_to_join > 0)
+    {
+        CUDA_TRY (cudaMemcpy(left_j_cols.data(), join_cols.get_column(0)->data, 
+                             cudf::size_of(join_cols.get_column(0)->dtype)*num_cols_to_join , cudaMemcpyDeviceToHost));
+    }
+    if (num_merging_join_cols > 0)
+    {
+        CUDA_TRY (cudaMemcpy(l_common_name_join_ind.data(), merging_join_cols.get_column(0)->data,
```

Comment:
Something is wrong here. The memory for `l_common_name_join_ind` is of size `sizeof(int) * num_merging_join_cols`. But this `cudaMemcpy is copying based on the `cudf::sizeof()` of the column's dtype.  What if the dtype is larger than int? We will get a CUDA error. I don't understand why you are copying rows of a column into a vector of indices.

Reply:
Same goes for all 3 of these `cudaMemcpy` calls.

Reply:
Also, why are we copying column data back to host anyway? Is there some of this processing we could just do in a kernel?

Reply:
As we need them for looping and other stuff throughout the function, using kernel will not be efficient.